File: python/781_森林中的兔子_medium.py
class Solution:
    def numRabbits(self, answers: List[int]) -> int:
        isSeen = dict()
        n, exAns = len(answers), 0
        for an in answers:
            if an == 0:
                continue
            elif an not in isSeen:
                isSeen[an] = an
                exAns += an
            else:
                if isSeen[an] > 0:
                    isSeen[an] -= 1
                    exAns -= 1
                else:
                    isSeen[an] = an
                    exAns += an
        return n + exAns

File: cpp/781_森林中的兔子_medium.cpp
class Solution {
public:
    int numRabbits(vector<int>& answers) {
        unordered_map<int, int> isSeen;
        int n = answers.size();
        int exAns = 0;
        for (auto an : answers) 
            if (an == 0)
                continue;
            else if (isSeen.find(an) == isSeen.end()) {
                isSeen[an] = an;
                exAns += an;
            }
            else {
                if (isSeen[an] > 0) {
                    isSeen[an]--;
                    exAns--;
                }
                else {
                    isSeen[an] = an;
                    exAns += an;
                }
            }
        return n + exAns;

    }
};